tests/environment.cc

```diff
@@ -545,6 +575,194 @@ test_groupIsLocked_upgrades()
+  manifestRaw.registry         = registryWithNixpkgs;
+
+  /* Create expected lockfile, reusing manifestRaw */
+  LockfileRaw lockfileRaw;
+  lockfileRaw.packages = { { _system, { { "hello", mockHelloLocked } } } };
+  lockfileRaw.manifest = manifestRaw;
+
+  Lockfile lockfile( lockfileRaw );
+
+  /* Name the group hello is in. */
+  ManifestRaw    modifiedManifestRaw( manifestRaw );
+  nlohmann::json helloJSON = { { "package-group", "blue" } };
+  modifiedManifestRaw.install->at( "hello" )
+    = ManifestDescriptorRaw( helloJSON );
+  EnvironmentManifest manifest( modifiedManifestRaw );
+
+  /* The locked input is returned by `getGroupInput`. */
+  TestEnvironment environment( std::nullopt, manifest, lockfile );
+  for ( const InstallDescriptors & group : manifest.getGroupedDescriptors() )
+    {
+      auto input = environment.getGroupInput( group, lockfile, _system );
+      EXPECT( input.has_value() );
+      EXPECT( equalLockedInputRaw( *input, mockHelloLocked.input ) );
+    }
+  return true;
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+/**
+ * @brief `getGroupInput` respects the lock of a package that used to be in a
+ *        group over a package that was just added.
+ */
+bool
+test_getGroupInput1()
+{
+  /* Create manifest with hello and curl in separate group */
+  ManifestRaw           manifestRaw;
+  nlohmann::json        curlJSON = { { "package-group", "blue" } };
+  ManifestDescriptorRaw curl( curlJSON );
+  manifestRaw.install = { { "hello", std::nullopt }, { "curl", curl } };
+  manifestRaw.options = Options {};
+  manifestRaw.options->systems = { _system };
+  manifestRaw.registry         = registryWithNixpkgs;
+
+  /* Create expected lockfile, reusing manifestRaw, and ensuring curl's input is
+   * different than hello's. */
+  LockedPackageRaw mockCurlLockedModified( mockCurlLocked );
+  mockCurlLockedModified.input.url = "not the same as hello's input";
+  LockfileRaw lockfileRaw;
+  lockfileRaw.packages = { { _system,
+                             { { "hello", mockHelloLocked },
+                               { "curl", mockCurlLockedModified } } } };
+  lockfileRaw.manifest = manifestRaw;
+
+  Lockfile lockfile( lockfileRaw );
+
+  /* Move hello to the same group as curl. */
+  ManifestRaw    modifiedManifestRaw( manifestRaw );
+  nlohmann::json helloJSON = { { "package-group", "blue" } };
+  modifiedManifestRaw.install->at( "hello" )
+    = ManifestDescriptorRaw( helloJSON );
+  EnvironmentManifest manifest( modifiedManifestRaw );
+
+  /* The locked input is returned by `getGroupInput`. */
+  TestEnvironment environment( std::nullopt, manifest, lockfile );
+  for ( const InstallDescriptors & group : manifest.getGroupedDescriptors() )
+    {
+      auto input = environment.getGroupInput( group, lockfile, _system );
+      EXPECT( input.has_value() );
+      // TODO this prints "Expectation failed" which should probably be silenced
```

Comment:
I can fix that if that's annoying, or if you have a better suggestion for how to write `equalLockedInputRaw` let me know @aakropotkin

Reply:
does `EXPECT( ( ! equalLockedInputRaw( ... ) ) );` with extra parens help at all?

It might just be macro weirdness